src/wallet/wallet2.cpp

```diff
@@ -4255,6 +4255,447 @@ static uint32_t get_count_above(const std::vector<wallet2::transfer_details> &tr
+      {
+        // Only add if new
+        auto confirmed_tx = m_confirmed_txs.find(tx_hash);
+        if(confirmed_tx == m_confirmed_txs.end()) {
+          // tx is added to m_unconfirmed_txs - move to confirmed
+          if(m_unconfirmed_txs.find(tx_hash) != m_unconfirmed_txs.end()) 
+          { 
+            process_unconfirmed(tx_hash, dummy_tx, t.height);
+          }
+          // Tx sent by another wallet instance
+          else
+          {
+            confirmed_transfer_details ctd;
+            ctd.m_amount_in = amount_sent;
+            ctd.m_amount_out = amount_sent;
+            ctd.m_change = 0;
+            ctd.m_payment_id = payment_id;
+            ctd.m_block_height = t.height;
+            ctd.m_timestamp = t.timestamp;
+            m_confirmed_txs.emplace(tx_hash,ctd);
+          }
+          if (0 != m_callback)
+          {
+            m_callback->on_money_spent(t.height, tx_hash, dummy_tx, amount_sent, dummy_tx);
+          } 
+        }
+        // If not new - check the amount and update if necessary.
+        // when sending a tx to same wallet the receiving amount has to be credited
+        else
+        {
+          if(confirmed_tx->second.m_amount_in != amount_sent || confirmed_tx->second.m_amount_out != amount_sent)
+          {
+            MDEBUG("Adjusting amount sent/received for tx: <" + t.hash + ">. Is tx sent to own wallet? " << print_money(amount_sent) << " != " << print_money(confirmed_tx->second.m_amount_in));
+            confirmed_tx->second.m_amount_in = amount_sent;
+            confirmed_tx->second.m_amount_out = amount_sent;
+            confirmed_tx->second.m_change = 0;
+          }
+        }
+      }
+    }    
+  }
+  // TODO: purge old unconfirmed_txs
+  remove_obsolete_pool_txs(pool_txs);
+
+  // Calculate wallet balance
+  m_light_wallet_balance = ires.total_received-wallet_total_sent;
+  // MyMonero doesnt send unlocked balance
+  if(ires.total_received_unlocked > 0)
+    m_light_wallet_unlocked_balance = ires.total_received_unlocked-wallet_total_sent;
+  else
+    m_light_wallet_unlocked_balance = m_light_wallet_balance;
+}
+
+bool wallet2::light_wallet_parse_rct_str(const std::string& rct_string, const crypto::public_key& tx_pub_key, uint64_t internal_output_index, rct::key& decrypted_mask, rct::key& rct_commit, bool decrypt) const
+{
+  // rct string is empty if output is non RCT
+  if (rct_string.empty())
+    return false;
+  // rct_string is a string with length 64+64+64 (<rct commit> + <encrypted mask> + <rct amount>)
+  rct::key encrypted_mask;
+  std::string rct_commit_str = rct_string.substr(0,64);
+  std::string encrypted_mask_str = rct_string.substr(64,64);
+  THROW_WALLET_EXCEPTION_IF(string_tools::validate_hex(64, rct_commit_str), error::wallet_internal_error, "Invalid rct commit hash: " + rct_commit_str);
+  THROW_WALLET_EXCEPTION_IF(string_tools::validate_hex(64, encrypted_mask_str), error::wallet_internal_error, "Invalid rct mask: " + encrypted_mask_str);
```

Comment:
For the record, don't trust the backend, as it might get MITM'd. Always check :P